src/Stratis.Bitcoin/P2P/Peer/NetworkPeer.cs

```diff
@@ -199,19 +200,55 @@ public bool IsConnected
             }
         }
 
+        public static IPEndPoint GetHandshakedEndPoint(INetworkPeer peer)
+        {
+            if (peer.Inbound && peer.State == NetworkPeerState.HandShaked)
+            {
+                IPEndPoint addressFrom = peer.PeerVersion?.AddressFrom.MapToIpv6();
+
+                if (addressFrom != null && !addressFrom.Address.Equals(IPAddress.IPv6Any))
+                {
+                    if (peer.PeerEndPoint.Address.IsRoutable(false) || !addressFrom.Address.IsRoutable(false))
+                        return new IPEndPoint(peer.PeerEndPoint.Address.EnsureIPv6(), addressFrom.Port);
+
+                    return addressFrom;
+                }
+            }
+
+            return peer.PeerEndPoint.MapToIpv6();
+        }
+
         /// <inheritdoc />
-        public bool MatchRemoteIPAddress(IPAddress ip, int? port = null)
+        public IPEndPoint GetHandshakedEndPoint()
+        {
+            return GetHandshakedEndPoint(this);
+        }
+
+        /// <inheritdoc />
+        public bool MatchLocalIPAddress(IPAddress ip, int? port = null)
         {
-            bool isConnectedOrHandShaked = (this.State == NetworkPeerState.Connected || this.State == NetworkPeerState.HandShaked);
+            IPAddress compareAddress = this.PeerEndPoint.Address.EnsureIPv6();
+            int? comparePort = this.PeerEndPoint.Port;
+
+            if (this.Inbound)
+            {
+                IPEndPoint addressFrom = this.PeerVersion?.AddressFrom.MapToIpv6();
+
+                if (addressFrom != null && !addressFrom.Address.Equals(IPAddress.IPv6Any))
+                    comparePort = addressFrom.Port;
+                else
+                    comparePort = null;
+            }
 
-            bool isAddressMatching = this.RemoteSocketAddress.Equals(ip)
-                                     && (!port.HasValue || port == this.RemoteSocketPort);
+            return compareAddress.Equals(ip.EnsureIPv6()) && (!port.HasValue || !comparePort.HasValue || port == comparePort);
+        }
 
-            bool isPeerVersionAddressMatching = this.PeerVersion?.AddressFrom != null
-                                                && this.PeerVersion.AddressFrom.Address.Equals(ip)
-                                                && (!port.HasValue || port == this.PeerVersion.AddressFrom.Port);
+        /// <inheritdoc />
+        public bool MatchRemoteIPAddress(IPAddress ip, int? port = null)
+        {
+            IPEndPoint compareTo = this.GetHandshakedEndPoint();
 
-            return (isConnectedOrHandShaked && isAddressMatching) || isPeerVersionAddressMatching;
+            return compareTo.Address.Equals(ip.EnsureIPv6()) && (!port.HasValue || port == compareTo.Port);
```

Comment:
The logic here seems to be different. In the previous code if the peer was not connected we checked the `PeerVersion`s `AddressFrom`. In other words the previous `MatchRemoteIPAddress` method would have returned true of the peer was NOT connected but the `this.PeerVersion.AddressFrom` matched the IP to check.

Here we compare against the peer's `PeerEndpoint` address which is returned from the new `GetHandshakedEndPoint` method.

Reply:
Also, the previous `MatchRemoteIPAddress` did not distinguish between inbound and outbound but now the new `GetHandshakedEndPoint` method does?

Reply:
Taking into account the type of address we are comparing against is the only way to perform the comparison correctly.

Say a local inbound peer has connected to us. That local inbound peer may also have a remote ip address discovered by itself connecting to remote peers. However we don't want the comparison to use that remote ip address if we only know the peer's local ip address -e.g: fed ip 127.0.0.1:1234. Here we choose to compare using `MatchLocalIPAddress`.

Conversely if a peer connects to us and we want to identify it using its remote ip address we use `MatchRemoteIPAddress'. 

The decision on which method to use happens in `FindByIp`.

I hope you can see that the legacy code which did not make this distinction would have to resort to all kind of tricks, involving comparing on both addressed or even ignoring port numbers. Ignoring port numbers have foreseeable issues in certain select circumstances..